# Linear Program Formulations
Based on Lab 9 of ENGRI 1101 Fall 2019

<font color='red'> **Comments in red** </font>

<font color='blue'> **Solutions in blue** </font>

**Objectives:**
- Practice in formulating linear programming problems
- Introduce the idea of linear relaxation
- Demonstrate the relationship between an integer program and its linear relaxation

**Brief description:**
   In this lab we will see examples of modeling various problems as Linear Programs (LPs), coding and solving them with OR-Tools, and start considering the question of what can be done when the desired output of a mathematical model should be integer.

**General Advice and troubleshooting:** Need to switch to OR-Tools troubleshooting

## Part 0: Preliminaries
This lab involves formulating two linear programs and solving them in OR-Tools. Formulating linear programs is tricky, and it is a skill best taught through blood, sweat, and tears. You are HIGHLY encouraged to work in small groups of 2-3 on this lab and to brainstorm and debate formulations together. If you’re unsure of how to formulate something, get help from your TA!

<font color='red'> Blood, sweat, and tears... </font>

## Part 1: A Work-scheduling problem
[Adapted from Hillier & Lieberman] A new company is being set up to take telephone orders. The phone lines are staffed around the clock. People are hired to work in 8-hour shifts, starting at 6AM, 8AM, 12PM, 4PM, or 10PM. From survey data, the management has concluded that there are minimum staffing requirements at certain times of the day, as given in the table below.

| Time period   | Minimum Staff Size |
|---------------|--------------------|
| 6AM-8AM       | 48                 |
| 8AM-10AM      | 79                 |
| 10AM-Noon     | 65                 |
| Noon-2PM      | 87                 |
| 2PM-4PM       | 64                 |
| 4PM-6PM       | 87                 |
| 6PM-8PM       | 64                 |
| 8PM-10PM      | 73                 |
| 10PM-Midnight | 82                 |
| Midnight-6AM  | 43                 |

A person is paid differently, depending on their assigned shift, where the daily salary is 170, 160, 175, 180, and 195, for the five shifts, respectively. The management would like to decide the number of people to be hired for each shift so as to minimize the total salary costs.
The first part of this lab is aimed at helping you to formulate this problem as a linear programming problem. As usual, the first step is to make a first attempt at stating the decision variables. Conveniently, you did this in the prelab already!

**Q1:** Let’s call the variables $x_1, x_2, x_3, x_4,$ and $x_5$. Explain what each of these variables represents.

**A:** <font color='blue'> See the table. </font>

| Variable      | Meaning                                |
|---------------|----------------------------------------|
| $x_1$         | # of people hired for 6AM - 2PM shift  |    
| $x_2$         | # of people hired for 8AM - 4PM shift  | 
| $x_3$         | # of people hired for 12PM - 8PM shift | 
| $x_4$         | # of people hired for 4PM - 12AM shift | 
| $x_5$         | # of people hired for 10PM - 6AM shift | 

**Q2:** In terms of these decision variables, express the objective function.

**A:** <font color='blue'> $\min [170x_1 + 160x_2 + 175x_3 + 180x_4 + 195x_5]$ </font>

**Q3:** In your own words (i.e., using sentences and not mathematical notation) briefly describe what conditions must be satisfied for a given assignment of values of your decision variables to correspond to a feasible solution.

**A:** <font color='blue'> The number of people assigned at a given time priod (all of the people whose shift contains that time period) must meet the required number of people to be staffed </font>

**Q4:** Now try to express these restrictions as linear constraints in your decision variables. As a start, consider the period from 6AM to 8AM. You are required to staff this period with at least 48 people. Which shifts work during this time? How should you write this as a constraint?

**A:** <font color='blue'> The only shift that works at this time is the 6AM - 2PM shift represented by $x_1$. Hence, the constraint is $$x_1 \geq 48$$  </font>

**Q5:** Now complete your linear programming formulation (write the rest of the constraints).

**A:** <font color='blue'> See the table. </font>

|                           |
|---------------------------|
| $x_1 \geq 48$             |
| $x_1 + x_2 \geq 79$       |
| $x_1 + x_2 \geq 65$       |
| $x_1 + x_2 + x_3 \geq 87$ |
| $x_2 + x_3 \geq 64$       |
| $x_3 + x_4 \geq 87$       |
| $x_3 + x_4 \geq 64$       |
| $x_4 \geq 73$             |
| $x_4 + x_5 \geq 82$       |
| $x_5 \geq 43$             |

**Q6:** If deleting a constraint from an LP does not create any new feasible solutions, then the constraint is called *redundant*. Why is your constraint for the 10PM-midnight shift redundant? Feel free to cross it out.

**A:** <font color='blue'> Because we already have the constraint that $x_4 \geq 73$ and $x_5 \geq 43$. It follows that $x_4 + x_5 \geq 116$. Hence, $x_4 + x_5 \geq 82$ is always satisifed and does not remove any feasible solutions. </font>

Let’s look at the model for this problem in OR-Tools.

In [1]:
# import OR-Tools
from ortools.linear_solver import pywraplp as OR

# define the model
mod = OR.Solver('work_sched', OR.Solver.CBC_MIXED_INTEGER_PROGRAMMING)

# decision variables
#x = {i: mod.NumVar(0, mod.infinity(), 'x_'+str(i)) for i in range(1,6)}
x = {i: mod.IntVar(0, mod.infinity(), 'x_'+str(i)) for i in range(1,6)}

# objective function
mod.Minimize(170*x[1] + 160*x[2] + 175*x[3] + 180*x[4] + 195*x[5])

# constraints
mod.Add(x[1]                             >= 48)
mod.Add(x[1] + x[2]                      >= 79)
mod.Add(x[1] + x[2]                      >= 65)
mod.Add(x[1] + x[2] + x[3]               >= 87)
mod.Add(       x[2] + x[3]               >= 64)
mod.Add(              x[3] + x[4]        >= 87)
mod.Add(              x[3] + x[4]        >= 67) 
mod.Add(                     x[4]        >= 73)
mod.Add(                     x[4] + x[5] >= 82) 
mod.Add(                            x[5] >= 43);
mod.Add(30*x[1] + 10*x[2] + 10*x[3] + 30*x[4] + 50*x[5] >= 6500); # mail order

Study the model and ensure that you understand how it corresponds to the model you came up with. Note that in Python, you can use # to comment out parts: everything that appears after the # on a line is ignored. Also note that the spacing in defining the constraints is not neccesary.

**Q7:** Did you write down the correct formulation (Ignore commented out lines)? If not, what did you do wrong?

**A:** <font color='blue'> Depends. </font>

Now you will solve this model using OR-Tools. Run the cell below.

In [2]:
# solve and print solution
mod.Solve()
print('Objective =', mod.Objective().Value())
print('Solution:')
for v in mod.variables():
    print(v.name(),':', v.solution_value())

Objective = 40495.0
Solution:
x_1 : 48.0
x_2 : 51.0
x_3 : 13.0
x_4 : 74.0
x_5 : 44.0


**Q8:** What is the objective function value of the optimal solution? What is the optimal solution (i.e., the values of the decision variables)?

**A:**  <font color='blue'> The objective value is 40,315 and the solution is $x = (48, 50, 14, 73, 43)$ </font>

Each staff member is also expected to take care of processing a number of mail-in orders when not on the telephone. Each person in the 6am, 8am, noon, 4pm and 10pm shifts are expected to handle 30, 10, 10, 30, and 50 orders, respectively. The staffing is required to be sufficient to handle 6500 orders. This constraint has been included in the model (the line labeled mail order) but it was commented out. Uncomment it now by removing the #, thereby adding the constraint to our model. Run the cell containing the model to update it. Get the new solution.

In [3]:
# solve and print solution
mod.Solve()
print('Objective =', mod.Objective().Value())
print('Solution:')
for v in mod.variables():
    print(v.name(),':', v.solution_value())

Objective = 40495.0
Solution:
x_1 : 48.0
x_2 : 51.0
x_3 : 13.0
x_4 : 74.0
x_5 : 44.0


**Q9:** Report the new solution below.

**A:**  <font color='blue'> The objective value is 40,447 and the solution is $x = (48, 50, 14, 73, 43.6)$ </font>

**Q10:** The solution to this model is not actually viable as a real-world hiring policy. Why not? Can you tweak the solution you found to give a viable (but not necessarily optimal) hiring plan? What is the total cost of your plan? 

**A:**  <font color='blue'> We can not hire $\frac{6}{10}$ of a person so this solution is not viable. Instead, we can round up and let $x_5 = 45$ to give a feasible solution with a total cost of 40,525. </font>

In effect, the issue in the previous question was that we have an integer programming (IP) problem, not a linear programming (LP) problem. In the model file, comment out the line ```x = {i: mod.NumVar...``` and uncomment the line ```x = {i: mod.IntVar...```. This turns our LP into an IP. Run the cell containing the model to update it. Get the new solution.

In [4]:
# solve and print solution
mod.Solve()
print('Objective =', mod.Objective().Value())
print('Solution:')
for v in mod.variables():
    print(v.name(),':', v.solution_value())

Objective = 40495.0
Solution:
x_1 : 48.0
x_2 : 51.0
x_3 : 13.0
x_4 : 74.0
x_5 : 44.0


**Q11:** Write down the optimal solution and its value below. How does the optimal value of the integer program compare to the optimal value of the linear programming problem, its so-called linear relaxation? How does the optimal IP solution compare with the solution you proposed above?

**A:**  <font color='blue'> The optimal solution is x = (48, 51, 13, 74, 44) with objective value 40,495 which is greater than it's linear program relaxation but less than the proprosed solution from rounding up.</font>

## Part 2: A Production Process Model
[from Winston] Sunco Oil has three different processes that can be used to manufacture various types of gasoline. Each process involves blending oils in the company’s catalytic cracker. 
- Running process 1 for an hour costs 5 dollars and requires 2 barrels of crude oil 1 and 3 barrels of crude oil 2. The output of running process 1 for an hour is 2 barrels of gas 1 and 1 barrel of gas 2. 
- Running process 2 for an hour costs 4 dollars and requires 1 barrel of crude 1 and 3 barrels of crude 2. The output from running process 2 for an hour is 3 barrels of gas 2. 
- Running process 3 for an hour costs 1 dollar and requires 2 barrels of crude 2 and 3 barrels of gas 2. The output from running process 3 for an hour is 2 barrels of gas 3. 

Each week, 200 barrels of crude 1, at 2 dollars/barrel, and 300 barrels of crude 2, at 3 dollars/barrel, may be purchased. All gas produced can be sold at the following per-barrel prices: gas 1, 9; gas 2, 10; gas 3, 24. The catalytic cracker can be used for only 95 hours each week. Sunco Oil wishes to devise a weekly production plan that maximizes the production’s profit (revenues minus costs).

**Q12:** The above paragraph is pretty dense. Draw a pictorial representation of the flow of materials in this problem: from crude oil, to processes, to barrels of gas for sale. It might help to have one picture for each process (e.g. one picture for process 1 indicating what materials it needs to run and what materials it spits out). You might also want a chart indicating how much each barrel of crude oil / each process costs, as well as how much we profit from each gas.

**A:** <font color='blue'> Something like the following diagram:</font>

![title](process_model_diagram.png)

You’ll spend the rest of lab thinking about how to formulate this problem as a linear program. There are many ways to do this, but the most straightforward is to use 8 decision variables. Three of these are $\text{proc}_1$,$\text{proc}_2$ and $\text{proc}_3$, the numbers of hours spent running processes 1, 2 and 3. 

**Q13:** You should also have decision variables $\text{crude}_1$, $\text{crude}_2$, $\text{gas}_1$, $\text{gas}_2$, and $\text{gas}_3$. What do $\text{crude}_i$ and $\text{gas}_j$ represent?